map_model/src/make/traffic_signals.rs

```diff
@@ -329,6 +332,426 @@ fn stage_per_road(map: &Map, i: IntersectionID) -> ControlTrafficSignal {
+                }
+
+                if let Some(m) = remove_movement(&r2, &r1, &mut right) {
+                    stage1.protected_movements.insert(m.clone());
+                } else if let Some(m) = remove_movement(&r2, &r1, &mut left) {
+                    stage1.protected_movements.insert(m.clone());
+                }
+
+                // add right turns
+                turns(&r1, &right).iter().for_each(|t| {
+                    stage1.protected_movements.insert(*t);
+                });
+                turns(&r2, &right).iter().for_each(|t| {
+                    stage1.protected_movements.insert(*t);
+                });
+
+                // add left turns
+                turns(&r1, &left).iter().for_each(|t| {
+                    stage1.yield_movements.insert(*t);
+                    stage2.protected_movements.insert(*t);
+                });
+                turns(&r2, &left).iter().for_each(|t| {
+                    stage1.yield_movements.insert(*t);
+                    stage2.protected_movements.insert(*t);
+                });
+
+                // add stages if they have protected turns
+                if !stage1.protected_movements.is_empty() {
+                    add_stage(&mut ts, stage1);
+                }
+                if !stage2.protected_movements.is_empty() {
+                    add_stage(&mut ts, stage2);
+                }
+            } else {
+                turns(&r1, &right).iter().for_each(|t| {
+                    stage1.protected_movements.insert(*t);
+                });
+                turns(&r1, &left).iter().for_each(|t| {
+                    stage1.protected_movements.insert(*t);
+                });
+                add_stage(&mut ts, stage1);
+            }
+        }
+        Some(ts)
+    }
+
+    fn add_stage(ts: &mut ControlTrafficSignal, stage: Stage) {
+        // ensure a duplicate isn't added
+        if ts.stages.iter().all(|s| *s != stage) {
+            ts.stages.push(stage)
+        }
+    }
+
+    fn turns(from: &RoadID, turns: &Vec<MovementID>) -> Vec<MovementID> {
+        turns
+            .iter()
+            .filter_map(|turn| {
+                if from.0 == turn.from.id.0 {
+                    Some(turn.clone())
+                } else {
+                    None
+                }
+            })
+            .collect::<Vec<_>>()
+    }
+
+    fn remove_movement(
+        from: &RoadID,
+        to: &RoadID,
+        turns: &mut Vec<MovementID>,
+    ) -> Option<MovementID> {
+        let result = if let Some(m) = turns
+            .iter()
+            .filter(|turn| from.0 == turn.from.id.0 && to.0 == turn.to.id.0)
+            .next()
+        {
+            Some(m.clone())
+        } else {
+            None
+        };
+        if result.is_some() {
+            turns.retain(|turn| from.0 != turn.from.id.0 || to.0 != turn.to.id.0);
+        }
+        result
+    }
+
+    fn three_way_three_stage(map: &Map, i: IntersectionID) -> Option<ControlTrafficSignal> {
```

Comment:
Is this and the method below a copy of the ones in the outermost scope of this module?

Reply:
They are not, but are derived from them and altered. The 3-way has N/S R protected and a stage for N/S L protected and E yield. It seemed easier to use the existing mechanisms, rather than create my own. Similar changes occurred for 4-way.

Reply:
Got it, thanks. This will be more clear when we split things into more submodules later.